File: backend/app/design_agent/autofixer.js
#!/usr/bin/env node
// Static AST autofixer — Node companion (P1-10).
//
// Reads a JSON payload from stdin, runs four deterministic static fixers,
// writes a JSON result to stdout. One-shot per file (no persistent process).
//
// Per AD22: this is STATIC analysis only — it parses the TSX with
// @babel/parser and inspects the AST. NO runtime instantiation, NO browser
// harness, NO live-server check. The agent self-corrects from the structured
// error feedback the Python wrapper forwards as a tool_result `is_error`.
//
// Known-good lists (shadcn registry, package allowlist, semantic tokens) are
// passed in via the payload `data` field (single source of truth lives in
// autofixer_data.py) — this script holds no hardcoded lists.
//
// Module resolution: @babel/parser is resolved via NODE_PATH, which the Python
// wrapper points at prototype-runtime/node_modules (the existing Node install
// from the P0 Vite pipeline). No backend-side Node install is introduced.

const parser = require("@babel/parser");
const path = require("path").posix;

function readStdin() {
  return new Promise((resolve, reject) => {
    let data = "";
    process.stdin.setEncoding("utf8");
    process.stdin.on("data", (chunk) => { data += chunk; });
    process.stdin.on("end", () => resolve(data));
    process.stdin.on("error", reject);
  });
}

// --- Fixer (d): JSX / TS syntax soundness -----------------------------------
// Returns the parsed AST, or throws (caught in main -> jsx-syntax error).
function parseOrThrow(content) {
  return parser.parse(content, {
    sourceType: "module",
    plugins: ["typescript", "jsx"],
    errorRecovery: false,
  });
}

// --- Fixer (a): hallucinated imports ----------------------------------------
// Cross-references every import against the prototype's virtual filesystem
// (relative + `@/` alias) and the dependency allowlist (bare packages).
function fixerImports(ast, filePath, vfsPaths, data) {
  const errors = [];
  const vfs = new Set(vfsPaths);
  const knownPackages = new Set(data.known_packages);
  const shadcnRegistry = new Set(data.shadcn_registry);
  const fromDir = path.dirname(filePath || "");

  const resolvesInVfs = (base) => {
    const candidates = [
      base, base + ".ts", base + ".tsx",
      base + "/index.ts", base + "/index.tsx",
    ];
    return candidates.some((c) => vfs.has(c));
  };

  for (const node of ast.program.body) {
    if (node.type !== "ImportDeclaration") continue;
    const src = node.source.value;
    const line = node.loc?.start?.line ?? null;

    if (src.startsWith(".")) {
      // Relative import — resolve against the importing file's directory.
      const resolved = path.normalize(path.join(fromDir, src));
      if (!resolvesInVfs(resolved)) {
        errors.push({
          fixer: "hallucinated-import",
          line,
          col: null,
          message: `Relative import '${src}' does not resolve in the prototype filesystem.`,
          suggestion: "Write the imported file first, or correct the path.",
        });
      }
    } else if (src.startsWith("@/components/ui/")) {
      // shadcn component import — validate against the installed registry.
      const name = src.slice("@/components/ui/".length).split("/")[0];
      if (!shadcnRegistry.has(name)) {
        errors.push({
          fixer: "shadcn-component",
          line,
          col: null,
          message: `'${name}' is not in the shadcn/ui registry.`,
          suggestion: `Available: ${data.shadcn_registry.slice(0, 8).join(", ")}, …`,
        });
      }
    } else if (src.startsWith("@/")) {
      // Project-local non-component import (`@/` aliases the `src/` root).
      const base = path.normalize("src/" + src.slice("@/".length));
      if (!resolvesInVfs(base)) {
        errors.push({
          fixer: "hallucinated-import",
          line,
          col: null,
          message: `Project import '${src}' does not resolve.`,
          suggestion: "Write the file first, or check the alias path.",
        });
      }
    } else {
      // Bare package — must be allowlisted (or a @radix-ui/* subpackage).
      const pkg = src.startsWith("@")
        ? src.split("/").slice(0, 2).join("/")
        : src.split("/")[0];
      if (!knownPackages.has(pkg) && !pkg.startsWith("@radix-ui/")) {
        errors.push({
          fixer: "hallucinated-import",
          line,
          col: null,
          message: `Package '${pkg}' is not in the prototype's dependency allowlist.`,
          suggestion: "Prototypes use React + Vite + Tailwind + shadcn/ui only.",
        });
      }
    }
  }
  return errors;
}

// --- Fixer (b): Tailwind class validation -----------------------------------
// Flags colour utilities whose colour segment is a shadcn semantic token
// (`bg-foreground`, `text-primary`, `bg-primary-100`) — these do not exist in
// vanilla Tailwind. Deliberately permissive elsewhere: arbitrary values
// (`bg-[#abc]`), real palette colours (`bg-slate-50`), and all structural
// utilities pass.
const COLOUR_PREFIXES = [
  "bg", "text", "border", "ring", "divide", "fill", "stroke",
  "from", "via", "to", "outline", "decoration", "accent", "caret",
  "placeholder", "shadow", "ring-offset",
];

function isSemanticColourHallucination(token, semanticTokens) {
  // Arbitrary value (`bg-[#abc]`, `p-[14px]`) is always allowed.
  if (/\[[^\]]+\]/.test(token)) return false;
  for (const prefix of COLOUR_PREFIXES) {
    if (!token.startsWith(prefix + "-")) continue;
    const remainder = token.slice(prefix.length + 1);
    // Strip an optional numeric shade suffix (`primary-100` -> `primary`).
    const colourName = remainder.replace(/-\d{1,3}$/, "");
    if (semanticTokens.has(colourName)) return true;
    // A colour-prefixed token matched but the colour segment isn't semantic —
    // treat as valid (permissive). Stop after the first prefix match.
    return false;
  }
  return false;
}

function fixerTailwind(ast, data) {
  const errors = [];
  const semanticTokens = new Set(data.semantic_tokens);

  function inspectClassName(node) {
    const tokens = node.value.value.split(/\s+/).filter(Boolean);
    for (const raw of tokens) {
      // Strip `!` important modifier and any variant prefixes (`md:`,
      // `hover:`, `dark:md:`) — validate only the leaf utility.
      const leaf = raw.replace(/^!/, "").split(":").pop();
      if (!leaf) continue;
      if (isSemanticColourHallucination(leaf, semanticTokens)) {
        errors.push({
          fixer: "tailwind-class",
          line: node.loc?.start?.line ?? null,
          col: null,
          message: `Tailwind class '${raw}' uses a shadcn semantic token that does not exist in vanilla Tailwind.`,
          suggestion: "Use a concrete palette colour (e.g. 'bg-slate-50') or an arbitrary value (e.g. 'bg-[#0f172a]').",
        });
      }
    }
  }

  function walk(node) {
    if (!node || typeof node !== "object") return;
    if (
      node.type === "JSXAttribute" &&
      node.name?.name === "className" &&
      node.value?.type === "StringLiteral"
    ) {
      inspectClassName(node);
    }
    for (const key of Object.keys(node)) {
      if (key === "loc" || key === "start" || key === "end") continue;
      const value = node[key];
      if (Array.isArray(value)) value.forEach(walk);
      else if (value && typeof value === "object" && value.type) walk(value);
    }
  }

  walk(ast.program);
  return errors;
}

async function main() {
  const raw = await readStdin();
  let payload;
  try {
    payload = JSON.parse(raw);
  } catch {
    // Unparseable payload — fail open (best-effort contract).
    process.stdout.write(JSON.stringify({ ok: true }));
    return;
  }

  const filePath = payload.file_path || "";
  const content = payload.content || "";
  const vfsPaths = payload.virtual_fs_paths || [];
  const data = payload.data || { shadcn_registry: [], known_packages: [], semantic_tokens: [] };

  // Fixer (d) first: a parse failure short-circuits the structural fixers.
  let ast;
  try {
    ast = parseOrThrow(content);
  } catch (e) {
    process.stdout.write(JSON.stringify({
      ok: false,
      errors: [{
        fixer: "jsx-syntax",
        line: e.loc?.line ?? null,
        col: e.loc?.column ?? null,
        message: `JSX/TS parse error: ${e.message}`,
        suggestion: null,
      }],
    }));
    return;
  }

  const errors = [
    ...fixerImports(ast, filePath, vfsPaths, data),
    ...fixerTailwind(ast, data),
  ];
  process.stdout.write(JSON.stringify(errors.length ? { ok: false, errors } : { ok: true }));
}

main().catch((e) => {
  process.stderr.write(String(e?.stack || e));
  process.exit(1);
});

File: backend/app/design_agent/autofixer.py
"""Static AST autofixer — invokes the Node @babel/parser companion (P1-10).

Per agent-build-research.md §2.4 (v0's deterministic post-stream autofixer is
"the biggest reliability boost in the ecosystem") + §5.2 (hallucinated imports
is the most common single failure mode). Per AD22: static analysis is
permitted; browser self-testing is NOT (no runtime feedback loop).

Four deterministic fixers (each sub-250ms target per file; the actual AST work
lives in the Node companion autofixer.js):
1. Hallucinated-import detection (against the prototype virtual_fs + allowlist)
2. Tailwind class validation (shadcn semantic-token detection)
3. shadcn component validation (against the installed registry)
4. JSX/TS syntax soundness (parse failure -> is_error)

Return shape:
  {"ok": True}                                          on clean validation
  {"ok": False, "errors": [{"fixer", "line", "col",     on validation failure
                            "message", "suggestion"}]}

Best-effort contract: on Node-missing / timeout / subprocess failure / invalid
JSON, returns {"ok": True} and logs ONE warning. A broken validator must never
block the agent loop — tsc/build errors downstream catch what the static
analyser missed.

Module resolution: @babel/parser lives in prototype-runtime/node_modules (the
existing P0 Vite-pipeline install). The subprocess is launched with NODE_PATH
pointed there, so no backend-side Node install is introduced (AD13 / no new
build tooling).
"""
from __future__ import annotations

import asyncio
import json
import logging
import os
from pathlib import Path
from typing import Any

from app.design_agent.autofixer_data import payload_data
from app.design_agent.build_env import scrubbed_node_env

logger = logging.getLogger(__name__)

_AUTOFIXER_JS = Path(__file__).with_suffix(".js")
# design_agent -> app -> backend -> repo root; prototype-runtime is a sibling
# of backend/ holding the @babel/parser install (P0-02 pin). This is the
# production default location for the @babel/parser resolution.
_REPO_ROOT = Path(__file__).resolve().parents[3]
_PROTOTYPE_RUNTIME_NODE_MODULES = _REPO_ROOT / "prototype-runtime" / "node_modules"
_NODE_BIN = os.environ.get("NODE_BIN", "node")
_SUBPROCESS_TIMEOUT_S = 8.0


def _node_modules_path() -> Path:
    """Directory placed on NODE_PATH so `require('@babel/parser')` resolves.

    Defaults to prototype-runtime/node_modules (the P0 Vite install) in
    production. Overridable via `DESIGN_AGENT_NODE_PATH` — used by backend CI
    to point at an isolated @babel/parser install (the CI image does not carry
    the prototype-runtime node_modules). Production behaviour is unchanged when
    the env var is unset."""
    override = os.environ.get("DESIGN_AGENT_NODE_PATH")
    return Path(override) if override else _PROTOTYPE_RUNTIME_NODE_MODULES


def _subprocess_env() -> dict[str, str]:
    """Secret-free env for the Node subprocess, with NODE_PATH pointing at the
    node_modules that holds @babel/parser so `require('@babel/parser')` resolves
    without a backend-side install. Prepends to any inherited NODE_PATH rather
    than clobbering it.

    The autofixer runs Node on agent-generated (user-influenced) code, so — like
    the Vite build — it must NOT inherit the backend's secrets. We build from an
    allowlisted base (see build_env) instead of `dict(os.environ)`."""
    node_modules = str(_node_modules_path())
    existing = os.environ.get("NODE_PATH")
    node_path = (
        f"{node_modules}{os.pathsep}{existing}" if existing else node_modules
    )
    return scrubbed_node_env({"NODE_PATH": node_path})


async def run(file_path: str, content: str, virtual_fs: dict[str, str]) -> dict[str, Any]:
    """Run the autofixer on a single emitted file. Returns the structured result.

    `virtual_fs` is the prototype's in-memory file map (path -> content); the
    hallucinated-import fixer cross-references imports against its keys. Only
    `.tsx`/`.ts` files are validated; anything else short-circuits to ok without
    spawning Node. The Node script reads the payload from stdin and writes the
    result to stdout (one-shot; no persistent process).
    """
    if not file_path.endswith((".tsx", ".ts")):
        return {"ok": True}

    payload = json.dumps({
        "file_path": file_path,
        "content": content,
        "virtual_fs_paths": list(virtual_fs.keys()),
        "data": payload_data(),
    })

    # Run via asyncio.to_thread + synchronous subprocess.run so the autofixer
    # works under uvicorn's WindowsSelectorEventLoopPolicy, which does not support
    # asyncio.create_subprocess_exec (raises NotImplementedError). Mirrors the
    # approach used by storage.vite_build (same constraint, same fix).
    try:
        import subprocess as _subprocess
        result = await asyncio.to_thread(
            _subprocess.run,
            [_NODE_BIN, str(_AUTOFIXER_JS)],
            input=payload.encode("utf-8"),
            capture_output=True,
            timeout=_SUBPROCESS_TIMEOUT_S,
            env=_subprocess_env(),
        )
    except FileNotFoundError:
        logger.warning("autofixer_node_missing bin=%s", _NODE_BIN)
        return {"ok": True}
    except _subprocess.TimeoutExpired:
        logger.warning("autofixer_timeout file_path=%s", file_path)
        return {"ok": True}
    except Exception as exc:
        logger.warning("autofixer_subprocess_error file_path=%s error_class=%s", file_path, type(exc).__name__)
        return {"ok": True}

    if result.returncode != 0:
        logger.warning(
            "autofixer_subprocess_failed file_path=%s rc=%s stderr=%s",
            file_path, result.returncode,
            result.stderr.decode("utf-8", errors="replace")[:200],
        )
        return {"ok": True}

    try:
        return json.loads(result.stdout.decode("utf-8"))
    except json.JSONDecodeError:
        logger.warning("autofixer_invalid_json file_path=%s", file_path)
        return {"ok": True}



def format_errors_for_agent(result: dict[str, Any]) -> str:
    """Render the autofixer result as plain-text feedback for the agent's
    tool_result block. On ok results returns a one-line pass message."""
    if result.get("ok"):
        return "Static analysis passed."
    lines = ["Static analysis failed. Fix each error and retry:"]
    for err in result.get("errors", []):
        loc = (
            f"line {err.get('line')}, col {err.get('col')}"
            if err.get("line")
            else "(no location)"
        )
        suggestion = f" Suggestion: {err.get('suggestion')}" if err.get("suggestion") else ""
        lines.append(f"  - [{err.get('fixer')}] {loc}: {err.get('message')}{suggestion}")
    return "\n".join(lines)
